Error message not displayed when trying to save new workflow with existing title

**Describe the bug**
When trying to save a new workflow with the title of an already existing workflow, an error message is added to the internal logging files. On the frontend however, no corresponding message is displayed to the user. Instead, the user is redirected to the worfklow list and it seems like the workflow has been saved correctly.

**To Reproduce**
Steps to reproduce the behavior:
1. Go to the "Workflows" tab on the "Projects" list page
2. Create new workflow
3. Add the title of an already existing workflow
4. Click save

**Expected behavior**
When the user tries to save a new workflow with the title of an existing workflow, an corresponding error message should be displayed and the browser should stay in the workflow editor instead of redirecting the user to the workflow list.

**Release**
3.7.x, 3.8.x, main branch

**Additional context**
Similar to #6281

(partial) stack trace from log file:
```
[ERROR] 2024-12-19 08:54:53.600 [http-nio-8080-exec-4] WorkflowForm - org.kitodo.data.database.exceptions.DAOException: org.hibernate.exception.ConstraintViolationException: could not execute statement
org.kitodo.data.exceptions.DataException: org.kitodo.data.database.exceptions.DAOException: org.hibernate.exception.ConstraintViolationException: could not execute statement
        at org.kitodo.production.services.data.base.SearchService.save(SearchService.java:302) ~[classes/:?]
        at org.kitodo.production.forms.WorkflowForm.saveWorkflow(WorkflowForm.java:331) ~[classes/:?]
        at org.kitodo.production.forms.WorkflowForm.saveAndRedirect(WorkflowForm.java:151) ~[classes/:?]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
```

--- a/Kitodo/src/main/java/org/kitodo/production/forms/WorkflowForm.java
+++ b/Kitodo/src/main/java/org/kitodo/production/forms/WorkflowForm.java
@@ -334,7 +334,7 @@ public class WorkflowForm extends BaseForm {
     }
 
     private void saveWorkflow() throws DataException {
-        ServiceManager.getWorkflowService().save(this.workflow, true);
+        ServiceManager.getWorkflowService().saveWorkflow(this.workflow);
     }
 
     /**

--- a/Kitodo/src/main/java/org/kitodo/production/forms/WorkflowForm.java
+++ b/Kitodo/src/main/java/org/kitodo/production/forms/WorkflowForm.java
@@ -90,7 +90,7 @@ public class WorkflowForm extends BaseForm {
     /**
      * Get list of workflow statues for select list.
      *
-     * @return list of SelectItem objects
+     * @return array of SelectItem objects
      */
     public WorkflowStatus[] getWorkflowStatuses() {
         return WorkflowStatus.values();
@@ -160,7 +160,10 @@ public class WorkflowForm extends BaseForm {
             } else {
                 return this.stayOnCurrentPage;
             }
-        } catch (IOException | DAOException | DataException e) {
+        } catch (DataException e) {
+            Helper.setErrorMessage(e.getLocalizedMessage(), logger, e);
+            return this.stayOnCurrentPage;
+        } catch (IOException | DAOException e) {
             Helper.setErrorMessage("errorDiagramFile", new Object[] {this.workflow.getTitle() }, logger, e);
             return this.stayOnCurrentPage;
         } catch (WorkflowException e) {
@@ -226,7 +229,11 @@ public class WorkflowForm extends BaseForm {
      */
     public void archive() {
         this.workflow.setStatus(WorkflowStatus.ARCHIVED);
-        saveWorkflow();
+        try {
+            saveWorkflow();
+        } catch (DataException e) {
+            Helper.setErrorMessage(e.getLocalizedMessage(), logger, e);
+        }
     }
 
     /**
@@ -326,12 +333,8 @@ public class WorkflowForm extends BaseForm {
         return xmlDiagramName;
     }
 
-    private void saveWorkflow() {
-        try {
-            ServiceManager.getWorkflowService().save(this.workflow, true);
-        } catch (DataException e) {
-            Helper.setErrorMessage(e.getLocalizedMessage(), logger, e);
-        }
+    private void saveWorkflow() throws DataException {
+        ServiceManager.getWorkflowService().save(this.workflow, true);
     }
 
     /**
@@ -535,7 +538,7 @@ public class WorkflowForm extends BaseForm {
     /**
      * Get language.
      *
-     * @return language of the currently logged in user
+     * @return language of the currently logged-in user
      */
     public String getLanguage() {
         return ServiceManager.getUserService().getCurrentUser().getLanguage();

--- a/Kitodo/src/main/java/org/kitodo/production/services/data/WorkflowService.java
+++ b/Kitodo/src/main/java/org/kitodo/production/services/data/WorkflowService.java
@@ -147,4 +147,28 @@ public class WorkflowService extends ClientSearchService<Workflow, WorkflowDTO, 
     public List<Workflow> getAllActiveWorkflows() {
         return dao.getAllActive();
     }
+
+    private List<Workflow> getByTitle(String title) {
+        return dao.getByQuery("FROM Workflow WHERE title = :title", Collections.singletonMap("title", title));
+    }
+
+    /**
+     * Save given workflow if it is an existing workflow (e.g. if it does have an ID)
+     * or if it is a new workflow and no workflow with the same name exists.
+     *
+     * @param workflow the object to save
+     *
+     * @throws DataException if the given workflow is a new workflow and a workflow with the same name already exists
+     */
+    public void saveWorkflow(Workflow workflow) throws DataException {
+        if (Objects.nonNull(workflow.getId())) {
+            save(workflow, true);
+        } else {
+            if (getByTitle(workflow.getTitle()).isEmpty()) {
+                save(workflow, true);
+            } else {
+                throw new DataException(Helper.getTranslation("duplicateWorkflowTitle", workflow.getTitle()));
+            }
+        }
+    }
 }

--- a/Kitodo/src/main/resources/messages/errors_de.properties
+++ b/Kitodo/src/main/resources/messages/errors_de.properties
@@ -45,6 +45,7 @@ docketNotFound=Die angegebene Datei konnte nicht gefunden werden.
 docketTitleDuplicated=Der Laufzettel mit den gleichen Titel existiert schon.
 docTypeNotFound=docType ''{0}'' wurde nicht im selektierten Regelsatz gefunden
 duplicatedTitles=Es wurden dublette Ausgabe-Bezeichnungen gefunden. Dies erzeugt gegebenenfalls dublette Vorgangstitel!
+duplicateWorkflowTitle=Ein Workflow mit dem Titel "{0}" existiert bereits.
 emptySourceFolder=Der Quellordner zur Bildgenerierung ist leer
 errorDataIncomplete=Unvollst\u00E4ndige Daten\:
 errorDatabaseReading=Fehler beim Datenbanklesen von ''{0}'' with ID {1}.

--- a/Kitodo/src/main/resources/messages/errors_en.properties
+++ b/Kitodo/src/main/resources/messages/errors_en.properties
@@ -46,6 +46,7 @@ docketNotFound=The specified file could not be found.
 docketTitleDuplicated=The docket with the same title exists already.
 docTypeNotFound=docType ''{0}'' could not be found in selected ruleset
 duplicatedTitles=Duplicate issue designations were found. This may produce duplicate process titles!
+duplicateWorkflowTitle=Workflow with title "{0}" already exists.
 emptySourceFolder=The source folder is empty
 errorDataIncomplete=Incomplete data\:
 errorDatabaseReading=Error on reading database for ''{0}'' with ID {1}.

--- a/Kitodo/src/test/java/org/kitodo/production/services/data/WorkflowServiceIT.java
+++ b/Kitodo/src/test/java/org/kitodo/production/services/data/WorkflowServiceIT.java
@@ -12,6 +12,7 @@
 package org.kitodo.production.services.data;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 import java.util.List;
@@ -22,11 +23,12 @@ import org.junit.jupiter.api.Test;
 import org.kitodo.MockDatabase;
 import org.kitodo.SecurityTestUtils;
 import org.kitodo.data.database.beans.Workflow;
+import org.kitodo.data.exceptions.DataException;
 import org.kitodo.production.services.ServiceManager;
 
 public class WorkflowServiceIT {
 
-    private WorkflowService workflowService = ServiceManager.getWorkflowService();
+    private final WorkflowService workflowService = ServiceManager.getWorkflowService();
 
     @BeforeAll
     public static void prepareDatabase() throws Exception {
@@ -61,5 +63,14 @@ public class WorkflowServiceIT {
         assertEquals(1, workflows.size(), "Workflows were not found in database!");
 
         SecurityTestUtils.cleanSecurityContext();
+    }
+
+    @Test
+    public void shouldNotSaveNewWorkflowWithExistingTitle() {
+        Workflow workflow = new Workflow("test");
+        DataException dataException = assertThrows(DataException.class,
+                () -> workflowService.saveWorkflow(workflow),
+                "Expected DataException to be thrown when saving a new workflow with an existing title");
+        assertEquals("Workflow with title 'test' already exists.", dataException.getMessage());
     }
 }

--- a/Kitodo/src/test/java/org/kitodo/production/services/data/WorkflowServiceIT.java
+++ b/Kitodo/src/test/java/org/kitodo/production/services/data/WorkflowServiceIT.java
@@ -24,6 +24,7 @@ import org.kitodo.MockDatabase;
 import org.kitodo.SecurityTestUtils;
 import org.kitodo.data.database.beans.Workflow;
 import org.kitodo.data.exceptions.DataException;
+import org.kitodo.production.helper.Helper;
 import org.kitodo.production.services.ServiceManager;
 
 public class WorkflowServiceIT {
@@ -68,9 +69,10 @@ public class WorkflowServiceIT {
     @Test
     public void shouldNotSaveNewWorkflowWithExistingTitle() {
         Workflow workflow = new Workflow("test");
+        String expectedExceptionMessage = Helper.getTranslation("duplicateWorkflowTitle", "test");
         DataException dataException = assertThrows(DataException.class,
                 () -> workflowService.saveWorkflow(workflow),
                 "Expected DataException to be thrown when saving a new workflow with an existing title");
-        assertEquals("Workflow with title 'test' already exists.", dataException.getMessage());
+        assertEquals(expectedExceptionMessage, dataException.getMessage());
     }
 }

